### PR TITLE
feat(quantic): Exposed clearFilters option in quantic searchboxes

### PR DESCRIPTION
**Pull request by SimonMilord**
[SFINT-5785](https://coveord.atlassian.net/browse/SFINT-5785)

## IN THIS PR:

- Added the `keepFiltersOnSearch` properties on the `quanticSearchBox` and `quanticStandaloneSearchBox` that defaults to false (since `clearFilters` is enabled by default). Setting this to `true` makes it possible to not clear the facets after triggering a search (see demo below)
- Added unit tests for when `keepFiltersOnSearch` is true/false

## DEMO:

https://github.com/user-attachments/assets/7dc8e431-4e2d-499b-9ac5-7e6e93d90699

[SFINT-5785]: https://coveord.atlassian.net/browse/SFINT-5785?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js [hunk 1]
```diff
@@ -66,6 +66,13 @@ export default class QuanticSearchBox extends LightningElement {
    * @defaultValue false
    */
   @api disableRecentQueries = false;
+  /**
+   * Whether to disable clearing all active query filters when the end user submits a new query from the search box.
+   * @api
+   * @type {boolean}
+   * @defaultValue false
+   */
+  @api disableClearFilters = false;
```
[the remark this answers is not in the document]
Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js:75]: I was following the same pattern we had for `disableRecentQueries`
what do you think of `keepFacetsOnSearch` ?

[the remark this answers is not in the document]
Comment (mmitiche) [reply packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js:75]: The thing is that it's not that facets that will be kept but the the filters selected in the facets.
cause the facets will always be there no matter what.

[the remark this answers is not in the document]
Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js:75]: I propose `keepFiltersOnSearch` then

File: packages/quantic/force-app/main/default/lwc/quanticStandaloneSearchBox/__tests__/quanticStandaloneSearchBox.test.js [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+  for (const [key, value] of Object.entries(options)) {
+    element[key] = value;
+  }
+  document.body.appendChild(element);
+  return element;
+}
+
+function prepareHeadlessState() {
+  // @ts-ignore
+  mockHeadlessLoader.loadDependencies = functionsMocks.loadDependencies;
```
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticStandaloneSearchBox/__tests__/quanticStandaloneSearchBox.test.js:58]: why is this needed?

File: packages/quantic/force-app/main/default/lwc/quanticStandaloneSearchBox/templates/standaloneSearchBox.html [hunk 3]
```diff
@@ -24,6 +24,7 @@
       without-submit-button={withoutSubmitButton}
       number-of-suggestions={numberOfSuggestions}
       textarea={textarea}
+      keep-filters-on-search={keepFiltersOnSearch}
```
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticStandaloneSearchBox/templates/standaloneSearchBox.html:27]: this is also something to cover in the tests:
```
      const searchBox = element.shadowRoot.querySelector(selectors.searchBox);

      expect(searchBox).not.toBeNull();
      expect(searchBox.keepFiltersOnSearch).toBe(expectedValue);
```

**Review by mmitiche — COMMENTED** [review]
good job 👍

couple more comments and there is the following issue that's occurring with the unit tests of the standalone search box that need to be fixed: https://github.com/coveo/ui-kit/actions/runs/11598185613/job/32293878598?pr=4606

Screenshot 2024-10-31 at 8 44 30 AM

File: packages/quantic/force-app/main/default/lwc/quanticStandaloneSearchBox/__tests__/quanticStandaloneSearchBox.test.js (hunk 2 above)
Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticStandaloneSearchBox/__tests__/quanticStandaloneSearchBox.test.js:58]: otherwise I was getting this error: 

image

File: packages/quantic/force-app/main/default/lwc/quanticStandaloneSearchBox/templates/standaloneSearchBox.html (hunk 3 above)
Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticStandaloneSearchBox/templates/standaloneSearchBox.html:27]: oh yeah good point!

**Comment by SimonMilord** [comment]
> good job 👍
> 
> couple more comments and there is the following issue that's occurring with the unit tests of the standalone search box that need to be fixed: https://github.com/coveo/ui-kit/actions/runs/11598185613/job/32293878598?pr=4606
> 
> Screenshot 2024-10-31 at 8 44 30 AM

will mute the console error

**Comment by mmitiche** [comment]
> > good job 👍
> > couple more comments and there is the following issue that's occurring with the unit tests of the standalone search box that need to be fixed: https://github.com/coveo/ui-kit/actions/runs/11598185613/job/32293878598?pr=4606
> > Screenshot 2024-10-31 at 8 44 30 AM
> 
> will mute the console error

It's important to figure out the WHY before muting or doing any actions. Because this did not appear with other tests, only with this one.

**Comment by SimonMilord** [comment]
> > > good job 👍
> > > couple more comments and there is the following issue that's occurring with the unit tests of the standalone search box that need to be fixed: https://github.com/coveo/ui-kit/actions/runs/11598185613/job/32293878598?pr=4606
> > > Screenshot 2024-10-31 at 8 44 30 AM
> > 
> > 
> > will mute the console error
> 
> It's important to figure out the WHY before muting or doing any actions. Because this did not appear with other tests, only with this one.

fixed that by mocking the `getHeadlessConfiguration` method

File: packages/quantic/force-app/main/default/lwc/quanticStandaloneSearchBox/__tests__/quanticStandaloneSearchBox.test.js [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+
+  afterEach(() => {
+    cleanup();
+  });
+
+  it('construct itself without throwing', () => {
+    expect(() => createTestComponent()).not.toThrow();
+  });
+
+  describe('when the current page reference changes', () => {
```
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticStandaloneSearchBox/__tests__/quanticStandaloneSearchBox.test.js:117]: To follow a more logical and chronological order, I would put this describe under the describe `controller initialization`

**Review by mmitiche — APPROVED** [review]
LGTM ✅

**Review by mikegron — APPROVED** [review]

File: packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js [hunk 5]
```diff
@@ -66,6 +66,13 @@ export default class QuanticSearchBox extends LightningElement {
    * @defaultValue false
    */
   @api disableRecentQueries = false;
+  /**
+   * Whether to keep all active query filters when the end user submits a new query from the search box.
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js:70]:
```suggestion
   * Whether to keep all active query filters, such as selected facets, when the end user submits a new query from the search box.
```

File: packages/quantic/force-app/main/default/lwc/quanticStandaloneSearchBox/__tests__/quanticStandaloneSearchBox.test.js [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+
+      expect(functionsMocks.subscribe).toHaveBeenCalledTimes(1);
+    });
+
+    describe('when the current page reference changes', () => {
+      beforeAll(() => {
+        Object.defineProperty(window, 'location', {
+          writable: true,
+          value: {href: nonStandaloneURL},
+        });
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticStandaloneSearchBox/__tests__/quanticStandaloneSearchBox.test.js:130]: Interesting, can you explain this? perhaps a comment and a reference where you found the doc for this?

Comment (erocheleau) [reply packages/quantic/force-app/main/default/lwc/quanticStandaloneSearchBox/__tests__/quanticStandaloneSearchBox.test.js:130]: Also why do you need this?

**Review by erocheleau — APPROVED** [review]
lgtm!

File: packages/quantic/force-app/main/default/lwc/quanticStandaloneSearchBox/__tests__/quanticStandaloneSearchBox.test.js (hunk 6 above)
Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticStandaloneSearchBox/__tests__/quanticStandaloneSearchBox.test.js:130]: this is because in order to test that the `keep-filters-on-search` is properly passed in the `c-quantic-search-box`, you need `isStandalone` to be false. 

The `isStandalone` property is set like this whenever the page ref changes:
`this.isStandalone = !window.location.href.includes(this.redirectUrl);`

When the the URL does not include the redirection URL. So this weird logic is to set a different URL in the window location and force the situation where the we render the `c-quantic-search-box`

So you need to set this property in order to be able to emit the wire event for the test:
`CurrentPageReference.emit({url: nonStandaloneURL});`

Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticStandaloneSearchBox/__tests__/quanticStandaloneSearchBox.test.js:130]: For your first comment, this is the ref I used: https://stackoverflow.com/questions/54021037/how-to-mock-window-location-href-with-jest-vuejs

Its to mock the window location in jest.
